test/run_test.py

```diff
@@ -412,6 +416,16 @@ def get_selected_tests(options):
 
     return selected_tests
 
+def find_test_file(test_name, test_directory):
+    test_file_name = '{}.py'.format(test_name)
+    for root, dirs, files in os.walk(test_directory):
+        for file in files:
+            if file == test_file_name:
+                abspath = os.path.join(root, file)
+                return os.path.relpath(abspath, test_directory)
+
+    raise Exception('Could not find test file for: {}'.format(test_name))
```

Comment:
Hmm, it's my bad for not being more explicit. When I mentioned that the previous code was not "how I would have written from scratch", what I had meant is that if I were writing this from scratch, I would have referred to tests by qualified name, so e.g., `jit/test_builtins` rather than just by the basename `test_builtins`. There are a few reasons why this makes more sense to me than searching for a file with this name (this patch) or hard-coding the path associated with the name (the previous patch), but the main reason is, usually when people introduce a directory structure, they imagine that it's OK for a file with the same name to live in different directories; e.g., `jit/test_basic` and `distributed/test_basic`. Both the old code, and this new code, violate that assumption: you must name every test uniquely, which could lead to weirdly redundant names like `jit/test_jit_basic`

Reply:
Let me address this in a follow up PR. We can add this functionality, although one question I had was what user behavior we would like to have for run_test.py. Currently you can run specific tests as follows: `run_test.py -i jit` or `run_test.py -i dist_optimizer_spawn`. Do we want to change this to be: `run_test.py -i test_jit` and `run_test.py -i distributed/rpc/test_dist_optimizer_spawn` (this option would modify existing behavior of run_test.py and I'm not sure if that breaks anything else). Or do we want something like this: `run_test.py -i jit` and `run_test.py -i distributed/rpc/dist_optimizer_spawn` ?

Reply:
Actually, looks like we need to fix this in the current PR itself since we have the following:

```
test/jit/test_data_parallel.py
test/distributed/test_data_parallel.py (this test wasn't enabled in CI, but is now with this PR)
```
I can update the PR based on which approach mentioned in my previous comment makes sense to you.

Reply:
I don't know. Honestly, I have always hated the fact that we needed a `run_test.py` script at all; why can't I just run the test files directly! (There are reasons why this can't be conveniently done.) There seem to be two conflicting design goals:

* I don't want to change the API for `run_test.py`, because if you do you're going to affect people who run tests this way (e.g., you're going to have to change the CI scripts at the very least)
* I don't want complicated name mangling logic for translating inputs in the test script (since you need to split the path, strip off `test_` basename, and then put it back together)

I think, on balance, I would rather keep `run_test.py` simple (no mangling logic) and force people to use the new API. I suppose you could add some BC logic (e.g., if you can't find the test, prepend `test_` again) to keep the old API working. I would like it even *more* if I could just get rid of `run_test.py` entirely for test scripts that don't need it.